src/libostree/ostree-sysroot-deploy.c

```diff
@@ -882,46 +882,190 @@ ostree_sysroot_write_origin_file (OstreeSysroot         *sysroot,
+ * initramfs there, so we need to look in /usr/lib/ostree-boot first.
  */
 static gboolean
-get_kernel_from_tree (int             deployment_dfd,
-                      int            *out_boot_dfd,
-                      char          **out_kernel_srcpath,
-                      char          **out_kernel_namever,
-                      char          **out_initramfs_srcpath,
-                      char          **out_initramfs_namever,
-                      char          **out_bootcsum,
-                      GCancellable   *cancellable,
-                      GError        **error)
+get_kernel_from_tree (int                  deployment_dfd,
+                      OstreeKernelLayout **out_layout,
+                      GCancellable        *cancellable,
+                      GError             **error)
 {
-  g_autofree char *ret_kernel_srcpath = NULL;
-  g_autofree char *ret_kernel_namever = NULL;
-  g_autofree char *ret_initramfs_srcpath = NULL;
-  g_autofree char *ret_initramfs_namever = NULL;
+  const char *legacy_paths[] = {"usr/lib/ostree-boot", "boot"};
   g_autofree char *kernel_checksum = NULL;
   g_autofree char *initramfs_checksum = NULL;
+  g_autoptr(OstreeKernelLayout) usrlib_modules_layout = NULL;
 
-  glnx_fd_close int ret_boot_dfd = glnx_opendirat_with_errno (deployment_dfd, "usr/lib/ostree-boot", TRUE);
-  if (ret_boot_dfd == -1)
+  /* First, gather from usr/lib/modules/$kver if it exists */
+  if (!get_kernel_from_tree_usrlib_modules (deployment_dfd, &usrlib_modules_layout, cancellable, error))
+    return FALSE;
+
+  /* Gather the legacy layout; if one exists, and it has an initramfs
+   * but the module layout doesn't, the legacy layout wins.  This is what happens
+   * with rpm-ostree with Fedora today, until rpm-ostree learns the new layout.
+   */
+  g_autoptr(OstreeKernelLayout) legacy_layout = g_new0 (OstreeKernelLayout, 1);
+  legacy_layout->boot_dfd = -1;
+  for (guint i = 0; i < G_N_ELEMENTS (legacy_paths); i++)
     {
-      if (errno != ENOENT)
-        return glnx_throw_errno_prefix (error, "%s", "openat(usr/lib/ostree-boot)");
+      const char *path = legacy_paths[i];
+      legacy_layout->boot_dfd = glnx_opendirat_with_errno (deployment_dfd, path, TRUE);
+      if (legacy_layout->boot_dfd == -1)
+        {
+          if (errno != ENOENT)
+            return glnx_throw_errno_prefix (error, "openat(%s)", path);
+        }
       else
         {
-          if (!glnx_opendirat (deployment_dfd, "boot", TRUE, &ret_boot_dfd, error))
-            return FALSE;
+          legacy_layout->initialized = TRUE;
+          break;
         }
     }
 
+  /* ret_boot_dfd will point to either /usr/lib/ostree-boot or /boot, let's
+   * inspect it.
+   */
   g_auto(GLnxDirFdIterator) dfditer = { 0, };
-  if (!glnx_dirfd_iterator_init_at (ret_boot_dfd, ".", FALSE, &dfditer, error))
+  if (!glnx_dirfd_iterator_init_at (legacy_layout->boot_dfd, ".", FALSE, &dfditer, error))
     return FALSE;
 
   while (TRUE)
```

Comment:
Could we factor this out as well into a separate function similar to `get_kernel_from_tree_usrlib_modules` that spits out the `legacy_layout`? Would make this function much nicer to read if it just has the business logic.